### ISSUE TITLE
ng update migration error within "migration-v9-undecorated-classes-with-di"

### ISSUE THREAD

**Issue by JohnnyDevNull** [angular/angular#34945]
# 🐞 Bug report

### Command (mark with an `x`)

- [ ] new
- [ ] build
- [ ] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [x ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc

### Description

My project setup is a angular workspace created with `--create-application=false` and then afterwards created a library and a application. The library is used also in other projects.

So the structure looks like:

* `projects/lvs-core` The library
* `projects/lvs-web` The application

Now i've tried to upgrade to the latest angular 9.rc-10 and got a lot of migration errors.

I've got hundreds of the following errors in the migrations for `migration-v9-undecorated-classes-with-di`:

* `Can't resolve all parameters`
* `Could not resolve @sk/lvs-core-dx relative to [object Object]..` 

### The Problem

I think the main problem is, that if the core library is not build the migration fails.
If i run `ng build lvs-core` and the try the migration again with `ng update @angular/core --migrate-only migration-v9-undecorated-classes-with-di --allow-dirty` it works because the library is now build under `dist/`, so i think it would be a good thing if the migration fails with a useful error to build the library first and then run the migration manually again?

You would have all neccessary information directly in the project... from the angular.json the libraries root path to check there the package.json for the libraries final name and with this match the path in the tsconfig.json to look for a existing build.

## 🌍 Your Environment
<pre><code>
     _                      _                 ____ _     ___ 
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | | 
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | | 
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 9.0.0-rc.10
Node: 12.13.0
OS: win32 x64

Angular: 9.0.0-rc.10
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router
Ivy Workspace: No

Package                            Version
------------------------------------------------------------    
@angular-devkit/architect          0.900.0-rc.10
@angular-devkit/build-angular      0.900.0-rc.10
@angular-devkit/build-ng-packagr   0.900.0-rc.10
@angular-devkit/build-optimizer    0.900.0-rc.10
@angular-devkit/build-webpack      0.900.0-rc.10
@angular-devkit/core               9.0.0-rc.10
@angular-devkit/schematics         9.0.0-rc.10
@angular/cdk                       8.2.3
@ngtools/webpack                   9.0.0-rc.10
@schematics/angular                8.3.23
@schematics/update                 0.900.0-rc.10
ng-packagr                         9.0.0-rc.7
rxjs                               6.5.4
typescript                         3.7.5
webpack                            4.41.2
</code></pre>

**Anything else relevant?**

root tsconfig.json
```
{
  "compileOnSave": false,
  "compilerOptions": {
    "baseUrl": "./",
    "downlevelIteration": true,
    "outDir": "./dist/out-tsc",
    "sourceMap": true,
    "declaration": false,
    "module": "esnext",
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "importHelpers": true,
    "target": "es2015",
    "typeRoots": [
      "node_modules/@types"
    ],
    "lib": [
      "es2018",
      "dom"
    ],
    "paths": {
      "@env/*": [
        "projects/lvs-web/src/environments/*"
      ],
      "@sk/lvs-core-dx": [
        "dist/lvs-core.build"
      ],
      "@sk/lvs-core-dx/*": [
        "dist/lvs-core.build/*"
      ],
      "jszip": [
        "node_modules/jszip/dist/jszip.min.js"
      ],
      "@app-shared": [
        "projects/lvs-web/src/app/shared/index"
      ],
      "@app-shared/*": [
        "projects/lvs-web/src/app/shared/*"
      ],
      "@app-services": [
        "projects/lvs-web/src/app/services/index"
      ],
      "@app-services/*": [
        "projects/lvs-web/src/app/services/*"
      ],
      "@app-interfaces": [
        "projects/lvs-web/src/app/interfaces/index"
      ],
      "@app-interfaces/*": [
        "projects/lvs-web/src/app/interfaces/*"
      ],
      "@app-components": [
        "projects/lvs-web/src/app/components/index"
      ],
      "@app-components/*": [
        "projects/lvs-web/src/app/components/*"
      ]
    },
    "resolveJsonModule": true,
    "esModuleInterop": true
  },
  "angularCompilerOptions": {
    "fullTemplateTypeCheck": true,
    "strictInjectionParameters": true,
    "enableIvy": false
  },
  "exclude": [
    "projects/lvs-schematics"
  ]
}

```

**Comment by devversion** [comment]
Thanks for creating this issue. We noticed a similar `ng update` limitation where pre-processors for templates (e.g. `pug`) are not properly handled.

It's true that we technically have the information about things that need to be built before we can instantiate the `AngularCompilerProgram`. Though this logic would seem quite convoluted for a single schematic. Also IIRC, the CLI doesn't leverage this information either (i.e. libraries still need to be built manually). 

The problem in general is that the `undecorated-classes-with-di` migration just accesses the standard `AngularCompilerProgram` from `@angular/compiler-cli`, while the CLI obviously requires libraries to be built beforehand, and has a [custom compiler program](https://github.com/angular/angular-cli/blob/master/packages/ngtools/webpack/src/angular_compiler_plugin.ts). We could use the same custom program in migrations and require libraries to be built beforehand, but it's unclear how this will work out from inside of schematics.

I agree that we can improve the error for this migration, and make it clear that libraries need to be built beforehand. In general though, this seems like a general conceptual problem for migrations that want to leverage the Angular compiler.

**Comment by devversion** [comment]
cc. @alan-agius4. Do you have any thoughts on this from the CLI team perspective?

**Comment by alan-agius4** [comment]
@devversion, indeed users always need to build their library manually prior to building the application there is no special handling for this in the `@ngtools/webpack` TypeScript compiler program. 

The `@ngtools/webpack` doesn't contain special handling in this regard and we don't have a custom compiler program https://github.com/angular/angular-cli/blob/0db3a699d461692a642f360cb63f699e04cd01ba/packages/ngtools/webpack/src/angular_compiler_plugin.ts#L361-L385 what is different is the compiler host, which uses Webpack's file system.

**Comment by devversion** [comment]
I looked into this again. There doesn't seem to be much we can do about this. The issue here is great as others that run into the same issue will be able to see what the solution is (i.e. building the library first)

Ideally though, we'd either have the possibility in the CLI to build libraries before (as explained in my comment above), or the path mappings would also point first to the actual sources. This seems to be the default in new CLI libraries. It looks like this is not the case here for `@sk/lvs-core-dx`.

https://github.com/angular/angular-cli/blob/master/packages/schematics/angular/library/index.ts#L218

Closing this as there is no action for this, and this migration is not going to run prospectively in v10 anymore. Also there is an obvious workaround, but nothing we can do to ease the migration.
